dali/operators/reader/loader/video/frames_decoder.cc

```diff
@@ -236,20 +243,84 @@ FramesDecoder::FramesDecoder(const char *memory_file, int memory_file_size, bool
   DetectVfr();
 }
 
+void FramesDecoder::CreateAvState(std::unique_ptr<AvState> &av_state, bool init_codecs) {
+    av_state->ctx_ = avformat_alloc_context();
+    DALI_ENFORCE(av_state_->ctx_, "Could not alloc avformat context");
+
+    uint8_t *av_io_buffer = static_cast<uint8_t *>(av_malloc(default_av_buffer_size));
+
+    AVIOContext *av_io_context = avio_alloc_context(
+      av_io_buffer,
+      default_av_buffer_size,
+      0,
+      &memory_video_file_.value(),
+      detail::read_memory_video_file,
+      nullptr,
+      detail::seek_memory_video_file);
+
+    av_state->ctx_->pb = av_io_context;
+
+    int ret = avformat_open_input(&av_state->ctx_, "", nullptr, nullptr);
+    DALI_ENFORCE(ret == 0, make_string("Failed to open video file ", Filename(), "due to ", detail::av_error_string(ret)));
+    av_state->stream_id_ = av_find_best_stream(av_state->ctx_, AVMEDIA_TYPE_VIDEO, -1, -1, nullptr, 0);
+    av_state->codec_params_ = av_state->ctx_->streams[av_state->stream_id_]->codecpar;
+
+    av_state->codec_ctx_ = avcodec_alloc_context3(av_state->codec_);
+    DALI_ENFORCE(av_state->codec_ctx_, "Could not alloc av codec context");
+
+    ret = avcodec_parameters_to_context(av_state->codec_ctx_, av_state->codec_params_);
+    DALI_ENFORCE(
+      ret >= 0,
+      make_string("Could not fill the codec based on parameters: ", detail::av_error_string(ret)));
+
+    av_state->packet_ = av_packet_alloc();
+    DALI_ENFORCE(av_state->packet_, "Could not allocate av packet");
+}
+
 void FramesDecoder::ParseNumFrames() {
   int curr_num_frames = 0;
-  while (av_read_frame(av_state_->ctx_, av_state_->packet_) >= 0) {
-    // We want to make sure that we call av_packet_unref in every iteration
-    auto packet = AVPacketScope(av_state_->packet_, av_packet_unref);
 
-    if (packet->stream_index != av_state_->stream_id_) {
-      continue;
+  if (IsFormatSeekable()) {
+    while (av_read_frame(av_state_->ctx_, av_state_->packet_) >= 0) {
+      // We want to make sure that we call av_packet_unref in every iteration
+      auto packet = AVPacketScope(av_state_->packet_, av_packet_unref);
+
+      if (packet->stream_index != av_state_->stream_id_) {
+        continue;
+      }
+      curr_num_frames++;
     }
-    curr_num_frames++;
+
+    num_frames_ = curr_num_frames;
+    Reset();
+  } else {
+    // Failover for unseekable video
+    auto current_position = memory_video_file_->position_;
+    memory_video_file_->Seek(0, SEEK_SET);
```

Comment:
Maybe we can copy `memory_video_file_` and pass this copy to `CreateAvState` ?

Reply:
Can we stay like this for now? I would like to decouple this even more in the next step. I just wanted to have minimal changes and the working decoding.

Reply:
Sure